active/0000-lazy-load-overview.md

```diff
@@ -0,0 +1,185 @@
+
+
+```
+app.import('vendor/dependency-1.js', { outputFile: 'assets/alternate-vendor.js'});
+```
+
+### Building multiple vendor JS assets from addons
+
+There is an [RFC](https://github.com/ember-cli/rfcs/pull/52) for this, but the concept is similar to the above, but
+extended for addons.
+
+### Building multiple JS assets for app code
+
+This isn't supported out of the box today, but it's possible to do. There are two example,
+[ember-cli-bundle-loader](https://github.com/MiguelMadero/ember-cli-bundle-loader) and
+[ember-lazy-loader](https://github.com/duizendnegen/ember-cli-lazy-load). The first one relies on a different file structure 
+and uses multiple ember-apps to build each asset. The latter relies on configuration to specify which app files will go to each
+asset.
+
+This RFC is limited to provide an overview of what is required, so this topic requires a bit more discusion on a
+separate RFC. TODO: create the separate RFC. Engines can provide a nice default, where we can have a convention of
+one asset per engine when lazyLoading is enabled, which can be overriden to allow for one or more engines per file.
+
+## Loading assets
+
+[ember-cli-bundle-loader](https://github.com/MiguelMadero/ember-cli-bundle-loader) provides a
+[service](https://github.com/MiguelMadero/ember-cli-bundle-loader/blob/master/addon/services/lazy-loader.js)
+to dynamically load CSS and JS. This service will be extracted to its own add-on and extended to support the
+asset metadata described below.
+
+## Lazy loading from components
+
+The service can be used by components when they need to load vendor JS or CSS assets. We have an example of this at
+Zenefits, where some of our components that depend on third-party libraries lazy-load them. In this particular case,
+we load the code on init and certain functions of the component are only available after the library is loaded.
+All of the code is removed to focus on the lazy-loading parts.
+
+```
+// app/components/z-table.js
+export default Ember.Component.extend({
+  lazyLoader: Ember.inject.service()
+  init() {
+    // This call is idempotent and the path will be mapped to a finger printed veresion
+  	this.get('lazyLoader').load('hands-on-table.js');
+  }
+});
+
+// app/components/z-table.hbs
+{{#if get(lazyLoader.loadedLibraries "hands-on-table.js"}}
+  {{!-- display other functionality that depends on the lazy-loaded lib --}}
+{{/if}}
+
+// ember-cli-build.js
+let app = new EmberApp(defaults, {});
+app.import('bower_components/hands-on-table/dist/hands-on-table.js', {outputFile: 'hands-on-table.js');
+```
+
+
+## Lazy Loading from routes
+
+For app code, we likely want an integration with the router, which can rely on this service to do the actual loading.
+
+When splitting the app's JS and CSS into multiple assets related to different sectionsof an application, we want to load the code dynamically. This today poses additional challenges because the handler (route) needs to exist to initiate a
```

Comment:
That's more or less the gist of it. If you return a `Promise` from `Router#getHandler`, any `Transition` should wait until it resolves with a handler. I picked up the work from @dgeb and with guidance from @rwjblue. Didn't see much reason for an RFC since this change should be totally transparent to Ember/Ember-CLI consumers.